linera-views/src/backends/lru_caching.rs

```diff
@@ -1,45 +1,305 @@
-static NUM_CACHE_FAULT: LazyLock<IntCounterVec> = LazyLock::new(|| {
-    prometheus_util::register_int_counter_vec("num_cache_fault", "Number of cache faults", &[])
-        .expect("Counter creation should not fail")
+/// The total number of value cache faults
+static NUM_CACHE_VALUE_FAULT: LazyLock<IntCounterVec> = LazyLock::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "num_cache_value_fault",
+        "Number of value cache faults",
+        &[],
+    )
+    .expect("Counter creation should not fail")
 });
 
 #[cfg(with_metrics)]
 /// The total number of cache successes
-static NUM_CACHE_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
-    prometheus_util::register_int_counter_vec("num_cache_success", "Number of cache success", &[])
-        .expect("Counter creation should not fail")
+static NUM_CACHE_VALUE_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "num_cache_value_success",
+        "Number of value cache success",
+        &[],
+    )
+    .expect("Counter creation should not fail")
+});
+
+#[cfg(with_metrics)]
+/// The total number of find cache faults
+static NUM_CACHE_FIND_FAULT: LazyLock<IntCounterVec> = LazyLock::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "num_cache_find_fault",
+        "Number of find cache faults",
+        &[],
+    )
+    .expect("Counter creation should not fail")
 });
 
-/// The `LruPrefixCache` stores the data for simple `read_values` queries
+#[cfg(with_metrics)]
+/// The total number of find cache successes
+static NUM_CACHE_FIND_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "num_cache_find_success",
+        "Number of find cache success",
+        &[],
+    )
+    .expect("Counter creation should not fail")
+});
+
+#[cfg(with_metrics)]
+/// Size of the inserted value entry
+static VALUE_CACHE_ENTRY_SIZE: LazyLock<HistogramVec> = LazyLock::new(|| {
+    prometheus_util::register_histogram_vec(
+        "value_cache_entry_size",
+        "Value cache entry size",
+        &[],
+        Some(vec![
+            10.0, 30.0, 100.0, 300.0, 1000.0, 3000.0, 10000.0, 30000.0, 100000.0, 300000.0,
+            1000000.0,
+        ]),
+    )
+    .expect("Histogram can be created")
+});
+
+#[cfg(with_metrics)]
+/// Size of the inserted find entry
+static FIND_CACHE_ENTRY_SIZE: LazyLock<HistogramVec> = LazyLock::new(|| {
+    prometheus_util::register_histogram_vec(
+        "find_cache_entry_size",
+        "Find cache entry size",
+        &[],
+        Some(vec![
+            10.0, 30.0, 100.0, 300.0, 1000.0, 3000.0, 10000.0, 30000.0, 100000.0, 300000.0,
+            1000000.0,
+        ]),
+    )
+    .expect("Histogram can be created")
+});
+
+#[derive(Clone, Hash, Eq, PartialEq, Ord, PartialOrd)]
+enum CacheEntry {
+    Find,
+    Value,
+}
+
+enum ValueCacheEntry {
+    DoesNotExist,
+    Exists,
+    Value(Vec<u8>),
```

Comment:
`Value(_)` already corresponds to `Exists` case. If you want to avoid allocations maybe it can be `Value(Cow<Vec<u8>>)` and let the caller decide whether to clone the `Vec` or not.

Reply:
In some cases we only check the database for existence of an entry, without loading it, so we won't have it in memory.

Reply:
That's a weird design choice. Every collection covers these functionalities but none chooses to mix it into a common variant. Maybe it's a sign that `Exists` shouldn't be here.

Reply:
Why? Isn't it better to cache any information we have about the entry?

Reply:
Can we make an assumption that if you make `db.contains(k)` then, at some point in the near future, you'll want to read it as well?

Reply:
No, that's not true.

(E.g. in some cases we only need to check that we have a certain data blob. If not, we need to request it. But we don't need to load the data right now.)

Reply:
Ok, you know better.